hashicorp/consul-k8s
# Present Consul server cert chain when using Vault secrets backend

Changes proposed in this PR:
- Present full cert chain (minus root cert) when using Vault as the secrets backend
- Update some bats tests to get them to pass

I was told by some HC support/CSM folks to open this as a PR. I don't know if this is the best way to accomplish this, or if the testing is sufficient, but hopefully this is useful. This was initially just meant to be a POC to show the engineering team what I did to address the issue on my end.

Currently, only the leaf cert gets presented by the Consul server (both UI and API). This works fine if you have a self-signed cert, but if you're using Vault as a secrets backend with an intermediate CA (and perhaps an offline root CA), then you can't currently verify the full chain.

The fix is to loop over the CA chain and append the intermediate cert(s) to the end of the server cert. This lets Consul present the full chain, which allows for proper validation.

This is related to #879, but only addresses the Vault secrets backend. Different work will need to be done to support the Kubernetes secrets backend.

How I've tested this PR:

I set up an offline root CA cert and used that to sign an L1 ICA in Vault, and used the L1 ICA to sign an L2 ICA. I then used that L2 ICA to issue certs for Consul. I added the root CA cert to my system's trust store and was able to access the UI in the browser without running into a TLS verification error.

It's been a while, but I'm pretty sure I also tested setting up a root CA directly in Vault and using that to issue certs for Consul.

Please feel free to suggest additional testing.

How I expect reviewers to test this PR:

- Deploy Vault to a local K8s cluster
- Set up a root CA in Vault
- Set up an L1 ICA in Vault signed by the root CA
- Optionally set up an L2 ICA (and so on)
- Configure a values.yaml for Consul that uses the Vault secrets backend for TLS certs with TLS enabled
- Either add the root CA cert to your system trust store and load the UI via a web browser, or use openssl to verify the chain via `openssl s_client -connect localhost:8501 -CAfile path/to/root-ca.pem`, which should return successfully
- Uninstall Consul and reinstall it to get certs issued directly from the Vault root CA, which shouldn't result in any extra certs being presented in the chain

Checklist:
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)
